### ISSUE TITLE
Can't synchronize to a remote instance using a certificate using alt. DNS Names

### ISSUE THREAD

**Issue by h122015** [MISP/MISP#1056]
I'm trying to synchronize to a remote instance running 2.4.17 but I realized yesterday that I couldn't connect to it due to the fact that it was using alternative DNS Names in the certificate. The connection test was failing with an "Unexpected error" and wasn't able to see the events on the remote instance with the Explore tool. I found a similar ticket (https://github.com/MISP/MISP/issues/581) and there it is stated that the solution is to upgrade PHP to version 5.6. I did that (current PHP version is 5.6.20) and the connection test ran OK and I could see the events on the remote instance through the Explore tool, so I assumed that the problem was solved.

Today I wanted to see which events I got from that remote instance but I couldn't find any and saw the following error messages in the job list:

stream_socket_client(): Peer certificate CN=`miranda.remote.org' did not match expected CN=`misp.remote.org' stream_socket_client(): Failed to enable crypto stream_socket_client(): unable to connect to ssl://misp.remote.org:443 (Unknown error) 

So basically the scheduled pulls/pushes haven't been working with this instance. I don't see any other errors in the logs

I really don't want to set ssl_verify_peer to False. Is there any other possibility to pull/push events from remote instances that use alternative DNS names?

**Comment by RichieB2B** [comment]
I think this is a CakePHP issue. It could be their code does not check for SSL SAN attributes.

**Comment by iglocska** [comment]
It could also be a PHP issue, we had this pop up at one point (https://github.com/MISP/MISP/issues/581) - which PHP version are you on?

**Comment by h122015** [comment]
5.6.20

**Comment by voycey** [comment]
I'm also seeing this recently as well on CakePHP with Amazon SES

```
stream_socket_client(): Peer certificate CN=`email-smtp.us-east-1.amazonaws.com' did not match expected CN=`ssl://email-smtp.us-east-1.amazonaws.com'
stream_socket_client(): Failed to enable crypto
stream_socket_client(): unable to connect to ssl://email-smtp.us-east-1.amazonaws.com:465 (Unknown error)

```

seems to not like the ssl:// prefix on the Mail Transport causing their cert to not match the CN defined in email.php

**Comment by RichieB2B** [comment]
That SMTP example is very different. Indeed the two strings don't match because one has the scheme ssl:// prepended. 

CakePHP is setting CN_match in lib/Cake/Network/CakeSocket.php The PHP documentation is not very clear if this context setting supports SAN: http://php.net/manual/en/context.ssl.php Looking at the source of PHP 5.6.20 however, ext/openssl/xp_ssl.c does use a matches_san_list() function to verify the peer_name (and CN_match). Are you sure misp.remote.org is actually in the SAN of the certificate? One way to check this is:

```
openssl s_client -connect misp.remote.org:443 </dev/null | openssl x509 -noout -text|grep DNS:
```

**Comment by h122015** [comment]
$ openssl s_client -connect misp.remote.org:443 </dev/null | openssl x509 -noout -text|grep DNS:
depth=1 C = BM, O = QuoVadis Limited, CN = QuoVadis Global SSL ICA G2
verify error:num=20:unable to get local issuer certificate
verify return:0
DONE
                DNS:miranda.remote.org, DNS:misp.remote.org

**Comment by RichieB2B** [comment]
Thanks, that looks ok to me. This is very weird. What CakePHP version are you using? 

```
git submodule status
```

**Comment by voycey** [comment]
Hey guys this is referenced here as well with a fix: https://github.com/cakephp/cakephp/issues/8604#issuecomment-207131151

**Comment by voycey** [comment]
Can confirm that updating Cake Core to 2.8 solved this issue for me:

Please ensure you read the Migration guide as there are a couple of changes to be aware of:

http://book.cakephp.org/2.0/en/appendices/2-8-migration-guide.html

**Comment by h122015** [comment]
The commit f4a1a32 and then doing 

`git submodule update`

seemed not to have solved my problem :(

**Comment by h122015** [comment]
Current CakePHP version after applying the patch is:

`# git submodule status
3a2414bc8c01687290245f186cfa72d29b99d09e PyMISP (v2.2-2-g3a2414b)
79db545b698cc23e784890a7398075d1212e29cc app/Lib/cakephp (2.8.3-7-g79db545)
+98e100500680a68ecc4887c4af8a4c3c440ef1c4 app/files/taxonomies (98e1005)
`

**Comment by amuehlem** [comment]
I had the same problem and tried to upgrade PHP to 5.6 (https://launchpad.net/~ondrej/+archive/ubuntu/php5-5.6), then I could sync without any problems.

**Comment by h122015** [comment]
Issue has been solved. For some reason it started working fine

**Comment by iglocska** [comment]
"It works, but why does it work?" :)

**Comment by h122015** [comment]
:) I did the update to 2.4.41 and restarted the server and all of the sudden it started working

**Comment by iglocska** [comment]
Excellent news!
